### PR TITLE
feat: generic session keys & session pallet migration of PR 1026

### LINKED ISSUE #1749: [Aura→BABE] PC-01 — Committee pallet two-deep look-ahead
**Workstream:** Partner-chains module (raised in midnight-node)  
**Part of:** Aura→BABE consensus migration — [MIP draft](https://github.com/midnightntwrk/midnight-improvement-proposals/blob/docs/mip-aura-to-babe-migration/mips/mip-xxx-aura-to-babe-migration.md)  
**First-pass sizing (to confirm with team):** 4–6 person-days

## Summary
Extend the committee pallet to compute and stage two epochs ahead instead of single-slot lookahead.

## Scope
- Compute the further-ahead committee within the finality margin
- Expose the staged committee to runtime authority management (RT-04)

## Acceptance criteria
- Further-ahead committee available within finality margin; feeds RT-04

---
_First-pass breakdown derived from the MIP Implementation section. Sizing and implementation plan to be reviewed and confirmed by the team._

### LINKED ISSUE #1759: [Aura→BABE] SPK-02 (spike) — Session-key carrier decision
**Workstream:** Spike / open design question (midnight-node)  
**Part of:** Aura→BABE consensus migration — [MIP draft](https://github.com/midnightntwrk/midnight-improvement-proposals/blob/docs/mip-aura-to-babe-migration/mips/mip-xxx-aura-to-babe-migration.md)  
**First-pass sizing (to confirm with team):** 2–3 person-days

## Summary
Decide the session-key carrier: phantom holder over the current `pallet_partner_chains_session` vs a plain generic session key (if generic-key work lands this quarter). Requirement is unchanged (registered, proven, usable keys at flip); the choice affects only sequencing.

## Scope
- Compare phantom-holder vs generic session key
- Recommend carrier and sequencing

## Acceptance criteria
- Decision recorded; RT-03 implementation approach unblocked

## Dependencies
- Blocks RT-03

---
_First-pass breakdown derived from the MIP Implementation section. Sizing and implementation plan to be reviewed and confirmed by the team._

### PR DESCRIPTION
**Pull request by Klapeyron**
# Overview

Integrates upstream Partner Chains changes for automatic session key registration ([#1026](https://github.com/input-output-hk/partner-chains/pull/1026)) and removal of `pallet-partner-chains-session` ([#1025](https://github.com/input-output-hk/partner-chains/pull/1025)), adapted for midnight-node's `polkadot-stable2603` workspace.
Partner Chains runtimes now use stock Substrate `pallet_session` directly. `pallet_session_validator_management` (`SessionCommitteeManagement`) implements `SessionManager` and `ShouldEndSession` inline (the old `PalletSessionSupport` wrapper and `pallet-partner-chains-session` crate are removed). Committee members' session keys are registered automatically at genesis and on each committee rotation via `pallet_session::SessionInterface::set_keys`, so block producers no longer need to call `set_keys` manually.
Midnight-node-specific wiring:
- Runtime migrates off `pallet-partner-chains-session` to a direct `pallet_session::Config` impl delegated to `SessionCommitteeManagement`.
- `pallet_session::historical` is enabled and wired (`Historical` pallet + `historical::Config`) so privileged `SessionInterface::set_keys` works under `polkadot-stable2603`'s ownership-proof checks (dispatching the `set_keys` extrinsic with an empty proof silently fails).
- `pallet_session` extrinsics are disabled via `#[runtime::disable_call]`; key registration is internal only.
- Genesis chain spec no longer duplicates session keys — registration happens through committee management genesis.
- The local `session_manager` module (`ValidatorManagementSessionManager`) is removed.
- Runtime `spec_version` bumped to `002_001_000`; metadata regenerated.

## 🗹 TODO before merging

- [ ] Ready

## 📌 Submission Checklist

- [ ] All commits are signed off (`git commit -s`) for the DCO
- [ ] Changes are backward-compatible (or flagged if breaking)
- [ ] Pull request description explains why the change is needed
- [ ] Self-reviewed the diff
- [ ] I have included a change file, or skipped for this reason: 
- [ ] If the changes introduce a new feature, I have bumped the node minor version
- [ ] Update documentation (if relevant)
- [ ] Updated AGENTS.md if build commands, architecture, or workflows changed
- [ ] No new todos introduced

## 🧪 Testing Evidence

Please describe any additional testing aside from CI:

- [ ] Additional tests are provided (if possible)

## 🔱 Fork Strategy

- [ ] Node Runtime Update
- [ ] Node Client Update
- [ ] Other: 
- [ ] N/A

### CODE REVIEW

File: partner-chains/toolkit/committee-selection/pallet/src/pallet_session_support.rs [hunk 1]
```diff
@@ -61,6 +67,53 @@ impl<T: crate::Config + pallet_session::Config> pallet_session::SessionManager<T
...
+}
+
+// Ensures that all accounts tied to new committee members exist by incrementing their
+// account provider counts. This is a necessary temporary solution, because we don't check
+// whether a block producer's account exists or not, when selecting them to a committee.
+// A proper solution would either be:
+// - increasing provider count for an account for as long as it is in the active committee
+//   and decreasing it afterwards, or
+// - considering account existence when selecting the committee
+// This will be addressed in later development.
```
Comment (LGLO) [inline partner-chains/toolkit/committee-selection/pallet/src/pallet_session_support.rs:104]: Should we have a ticket for it? Was it fixed in PC?

File: partner-chains/toolkit/partner-chains-cli/src/create_chain_spec/mod.rs [hunk 2]
```diff
@@ -144,21 +144,25 @@ impl<Keys: MaybeFromCandidateKeys> CreateChainSpecConfig<Keys> {
 		}
 	}
 
-	/// Returns [pallet_partner_chains_session::GenesisConfig] derived from the config, using initial permissioned candidates
+	/// Returns [pallet_session::GenesisConfig] derived from the config, using initial permissioned candidates
 	/// as initial validators
-	pub fn pallet_partner_chains_session_config<T: pallet_partner_chains_session::Config>(
+	pub fn pallet_partner_chains_session_config<T: pallet_session::Config>(
```
Comment (LGLO) [inline partner-chains/toolkit/partner-chains-cli/src/create_chain_spec/mod.rs]: I think we should rename it this time. `pallet_session_config`

File: partner-chains/toolkit/partner-chains-cli/Cargo.toml [hunk 3]
```diff
@@ -42,9 +42,10 @@ sp-session-validator-management = { workspace = true, features = ["std"] }
 pallet-session-validator-management = { workspace = true, features = ["std"] }
 pallet-partner-chains-bridge = { workspace = true, features = ["std"] }
 sp-partner-chains-bridge = { workspace = true, features = ["std"] }
-pallet-partner-chains-session = { workspace = true, features = ["std"] }
+pallet-session = { workspace = true, features = ["std"] }
 pallet-sidechain = { workspace = true, features = ["std"] }
 sidechain-slots = { workspace = true }
+sp-staking = { workspace = true }
```
Comment (LGLO) [inline partner-chains/toolkit/partner-chains-cli/Cargo.toml]: Don't add it. Unused.

File: partner-chains/toolkit/partner-chains-cli/Cargo.toml [hunk 4]
```diff
@@ -61,7 +62,7 @@ sp-runtime = { workspace = true }
 [features]
 default = ["std"]
 std = []
-runtime-benchmarks = []
+runtime-benchmarks = ["pallet-partner-chains-bridge/runtime-benchmarks"]
```
Comment (LGLO) [inline partner-chains/toolkit/partner-chains-cli/Cargo.toml]: I don't understand this one. Why `partner-chains-cli` have runtime-benchmarks in the first place?
Please remove it.

File: partner-chains/toolkit/committee-selection/pallet/src/pallet_session_support.rs (hunk 1 above)
Comment (Klapeyron) [reply partner-chains/toolkit/committee-selection/pallet/src/pallet_session_support.rs:104]: https://github.com/midnightntwrk/midnight-node/issues/1848

File: partner-chains/changelog.md [hunk 5]
```diff
@@ -7,9 +7,16 @@ This changelog is based on [Keep A Changelog](https://keepachangelog.com/en/1.1.
 ## Changed
 
 * Updated polkadot-sdk dependency to polkadot-stable2603.
+* Updated partner-chains-smart-contracts (raw-scripts) dependency to v8.2.0. Not breaking.
```
Comment (LGLO) [inline partner-chains/changelog.md:10]: I think it is redundant

File: partner-chains/toolkit/committee-selection/pallet/src/pallet_session_support.rs [hunk 6]
```diff
@@ -1,68 +1,121 @@
-//! Implements Substrate's [pallet_session].
+//! Implements [pallet_session::SessionManager] and [pallet_session::ShouldEndSession] for [crate::Pallet].
 //!
 //! This implementation has lag of one additional PC epoch when applying committees to sessions.
```
Comment (LGLO) [inline partner-chains/toolkit/committee-selection/pallet/src/pallet_session_support.rs]: Is this comment still true? I isn't there immediate session rotation at the next block?

Comment (Klapeyron) [reply partner-chains/toolkit/committee-selection/pallet/src/pallet_session_support.rs]: It is part of https://github.com/midnightntwrk/midnight-node/pull/1801/changes/9563f13f80963ac72b8a511ac3a971ef1720a3e0, I removed it with https://github.com/midnightntwrk/midnight-node/pull/1800/commits/3464eceacf7c924828f86a93b110df7d8dccdead

**Review by LGLO — DISMISSED** [review]
Just some minor comments regarding leaving comments that looks obsolete.

**Comment by m2ux** [comment]
## PR Review Summary

**PR**: #1800 - Generic session keys & session pallet migration (fork → stock `pallet_session`)
**Reviewers**: [Code Review Agent](https://github.com/m2ux/workflow-server/blob/workflows/work-package/techniques/review-code.md) · [Test Suite Review Agent](https://github.com/m2ux/workflow-server/blob/workflows/work-package/techniques/review-test-suite.md) · [Validation Agent](https://github.com/m2ux/workflow-server/blob/workflows/work-package/activities/11-validate.yaml) · [Strategic Review Agent](https://github.com/m2ux/workflow-server/blob/workflows/work-package/activities/12-strategic-review.yaml)
**Reports**: [Prior Feedback Triage](https://github.com/shieldedtech/midnight-agent-eng/blob/main/.engineering/artifacts/planning/2026-07-11-review-midnight-node-pr-1800/01-prior-feedback-triage.md) · [Code Review](https://github.com/shieldedtech/midnight-agent-eng/blob/main/.engineering/artifacts/planning/2026-07-11-review-midnight-node-pr-1800/10-code-review.md) · [Test Suite Review](https://github.com/shieldedtech/midnight-agent-eng/blob/main/.engineering/artifacts/planning/2026-07-11-review-midnight-node-pr-1800/10-test-suite-review.md) · [Validation](https://github.com/shieldedtech/midnight-agent-eng/blob/main/.engineering/artifacts/planning/2026-07-11-review-midnight-node-pr-1800/11-validation-results.md)
**Date**: 2026-07-12

### Executive Summary

The migration off the forked partner-chains session pallet onto stock pallet_session is architecturally sound and lean — deletion-dominated at net −644 lines, the fork removed cleanly, and the head-commit revert of the additional-session acceleration is a deliberate, correct restoration of the session-index-to-epoch invariant. But the change ships two consensus-critical defects on paths that no genesis-started environment ever exercises: there is no live-chain storage migration despite reusing the fork's Session pallet prefix, and key-registration failure is logged and swallowed rather than aborting rotation. Both were raised in prior review and remain unaddressed at head, and all six blocking CI checks currently fail.

**Overall Rating**: Request Changes

---

### Prior Feedback Triage

Disposition of every prior comment and review on the PR (human and bot), determined before independent analysis at head `4e48f8db`. Full detail in the [Prior Feedback Triage report](https://github.com/shieldedtech/midnight-agent-eng/blob/main/.engineering/artifacts/planning/2026-07-11-review-midnight-node-pr-1800/01-prior-feedback-triage.md).

| # | Prior Comment | Author | Disposition | Reasoning |
|---|---------------|--------|-------------|-----------|
| [1](https://github.com/midnightntwrk/midnight-node/pull/1800#issuecomment-3592881551) | Pipeline job failures (6 checks) | datadog (bot) | Confirmed | 6 blocking checks still fail on head — caps the rating |
| [6](https://github.com/midnightntwrk/midnight-node/pull/1800#discussion_r3542134760) | Migrate session storage before switching pallet implementations | codex (bot) | Confirmed | Unaddressed; verified as CR-1 below — caps the rating |
| [9](https://github.com/midnightntwrk/midnight-node/pull/1800#discussion_r3544118330) | Key-registration failure only logs and continues | codex (bot) | Confirmed | Unaddressed; verified as CR-2 below — caps the rating |
| [7](https://github.com/midnightntwrk/midnight-node/pull/1800#discussion_r3543872411) | Changelog line looks redundant | LGLO (human) | Confirmed | Open; verified as CR-3 below |
| [2](https://github.com/midnightntwrk/midnight-node/pull/1800#discussion_r3510846388) | Ticket for the fix / fixed in PC? | LGLO (human) | Superseded | Tracking issue [#1848](https://github.com/midnightntwrk/midnight-node/issues/1848) created; resolved |
| [3](https://github.com/midnightntwrk/midnight-node/pull/1800#discussion_r3510856996) | Rename to pallet_session_config | LGLO (human) | Superseded | Lines changed by later commits; marked outdated |
| [4](https://github.com/midnightntwrk/midnight-node/pull/1800#discussion_r3510874936) | Unused dependency — don't add | LGLO (human) | Superseded | Addition removed by later commits |
| [5](https://github.com/midnightntwrk/midnight-node/pull/1800#discussion_r3510878975) | Remove runtime-benchmarks feature | LGLO (human) | Superseded | Removed by later commits |
| [8](https://github.com/midnightntwrk/midnight-node/pull/1800#discussion_r3543943994) | Is the one-epoch-lag doc still true? | LGLO (human) | Superseded | Acceleration reverted; the lag comment is accurate at head |
| [10](https://github.com/midnightntwrk/midnight-node/pull/1800#pullrequestreview-3542160658) | Approving review with minor comments | LGLO (human) | Confirmed | Human approval on record; residual nits are rows 7–8 |

---

### Code Review Findings

| # | Finding | Severity |
|---|---------|----------|
| [1](https://github.com/midnightntwrk/midnight-node/blob/4e48f8db68982028ae9df7bf27d4575808adf984/runtime/src/lib.rs#L1025-L1027) | No live-chain session-storage migration despite reusing the fork's Session prefix / pallet index 30 | Critical |
| [2](https://github.com/midnightntwrk/midnight-node/blob/4e48f8db68982028ae9df7bf27d4575808adf984/partner-chains/toolkit/committee-selection/pallet/src/pallet_session_support.rs#L80-L86) | Key registration logs and continues on failure, so the consensus set can silently diverge from the stored committee | Major |
| [3](https://github.com/midnightntwrk/midnight-node/blob/4e48f8db68982028ae9df7bf27d4575808adf984/partner-chains/changelog.md) | Stale changelog entry describes reverted acceleration behaviour | Minor |
| [4](https://github.com/midnightntwrk/midnight-node/blob/4e48f8db68982028ae9df7bf27d4575808adf984/partner-chains/toolkit/committee-selection/pallet/src/pallet_session_support.rs#L100-L103) | Provider count incremented on committee entry but never decremented (bounded leak, acknowledged) | Nit |

<details>
<summary>Finding Details</summary>

#### CR-1. No live-chain session-storage migration (Critical)
Stock pallet_session is registered as Session at pallet index 30, exactly the index and name the forked pallet held in the base. Storage prefixes derive from the pallet name, so on a live upgrade stock pallet_session inherits the fork's existing Session-prefixed storage. Validators, CurrentIndex, and DisabledValidators survive, but QueuedKeys, NextKeys, and KeyOwner never existed under the fork and read back as their empty ValueQuery defaults. No migration reconciles this — neither migration tuple contains a pallet_session entry and the migrations module is untouched by the diff. On the first post-upgrade session rotation, pallet_session applies an empty queued set and the Aura and GRANDPA authorities empty out one session after the upgrade, halting block production on any chain upgraded from fork-era state. The ValueQuery defaults launder the missing storage into a legal empty value, so nothing panics — the failure is silent and invisible to every genesis-started environment, which is why all tests, mocks, local nets, and CI pass. This is prior-feedback row 6, unresolved and unrebutted at head.
**Suggestion**: Add an OnRuntimeUpgrade migration for pallet_session that populates QueuedKeys, NextKeys, and KeyOwner from the pre-upgrade committee state before the first post-upgrade rotation. Keep the migration minimal and verify it with try-runtime, since the KeyOwner translation is itself high-consequence, untested code.

#### CR-2. Registration failure logs and continues (Major)
Per committee member, set_keys is called and on error the code logs at info and continues. Meanwhile the session hooks return the full committee account list unconditionally, independent of which members' registration succeeded. A member whose registration fails is dropped from the applied queued keys while still being reported to pallet_session as an active validator account, so the consensus authority set diverges from the committee the management pallet believes is active. The dedup guard is a second silent-drop path: if two distinct members present the same session key, the second gets a duplicate-key error and is dropped via the same info branch. If enough members drop, the committee shrinks below the finality threshold and finality stalls, with only info logs as evidence. The forked pallet treated registration failure as loud and fatal; this migration weakens it to log-and-continue. This is prior-feedback row 9, unresolved and unrebutted at head.
**Suggestion**: Abort the rotation (propagate the error / return no new set) or validate the full key set before applying any, so the applied set and the reported set cannot diverge.

#### CR-3. Stale changelog entry (Minor)
The changelog adds, under Changed, a line stating the session-manager and should-end-session implementations rotate one additional session to make pallet_session use authorities with less delay. That acceleration is the mechanism the head commit reverted; at head the code applies the committee with a one-epoch lag, not with less delay. The line promises behaviour the code does not deliver. This is prior-feedback row 7.
**Suggestion**: Correct the line to describe the one-epoch-lag behaviour actually delivered at head.

#### CR-4. Provider count never decremented (Nit)
The provide-committee-accounts path increments the frame-system provider count for new committee accounts with no matching decrement when an account leaves the committee — a bounded provider-reference leak. It is explicitly acknowledged in the code as a temporary solution with the proper fix deferred, and a tracking issue [#1848](https://github.com/midnightntwrk/midnight-node/issues/1848) exists. Bounded in practice by the finite validator-account population, hence a nit rather than higher. The proper fix (decrement on committee exit) trades this bounded leak for account-reaping races, so the temporary approach is defensible.
**Suggestion**: Track the decrement fix to closure under [#1848](https://github.com/midnightntwrk/midnight-node/issues/1848).

</details>

---

### Test Review Findings

| # | Gap | Severity |
|---|-----|----------|
| [1](https://github.com/midnightntwrk/midnight-node/blob/4e48f8db68982028ae9df7bf27d4575808adf984/partner-chains/toolkit/committee-selection/pallet/src/pallet_session_support.rs) | No test for the live-upgrade session-storage path (masks CR-1) | Critical |
| [2](https://github.com/midnightntwrk/midnight-node/blob/4e48f8db68982028ae9df7bf27d4575808adf984/partner-chains/toolkit/committee-selection/pallet/src/pallet_session_support.rs) | No test for the registration-failure rotation path (masks CR-2); mock cannot force a set_keys failure | Major |
| [3](https://github.com/midnightntwrk/midnight-node/blob/4e48f8db68982028ae9df7bf27d4575808adf984/partner-chains/toolkit/committee-selection/pallet/src/pallet_session_support.rs) | No test for the dedup / duplicate-key drop branch | Minor |

<details>
<summary>Finding Details</summary>

The two consensus-critical paths this PR introduces have zero test coverage, and the test harness structurally cannot express either. Every test constructs state at genesis, so every test passes while the defects live only on the non-genesis and failure paths.

#### TR-1. No live-upgrade storage test (Critical)
Nothing constructs storage in the fork's pre-upgrade layout (populated Validators and CurrentIndex, absent QueuedKeys, NextKeys, KeyOwner) and asserts the first post-upgrade rotation yields a non-empty queued validator set. Such a test would fail today, demonstrating the missing migration.
**Suggestion**: Add a migration test over the pre-upgrade storage layout asserting a non-empty queued set after the first post-upgrade rotation.

#### TR-2. No registration-failure test; harness cannot express it (Major)
No test drives a rotation where one member's set_keys fails. The mock seeds keys through a stock genesis state and provides no way to force set_keys to return an error, so the failure branch is unreachable under any current test — the harness pins the single all-succeed instance and hides the divergence path. Closing this requires extending the mock to inject a failure (e.g. a pre-registered conflicting key), not merely adding an assertion.
**Suggestion**: Extend the mock to inject a set_keys failure, then assert the rotation aborts or the reported set does not exceed the applied queued set.

#### TR-3. No dedup / duplicate-key test (Minor)
The batch dedup guard and the duplicate-key branch have no coverage.
**Suggestion**: Add a test asserting the intended (not silently-lossy) behaviour of the dedup and duplicate-key paths.

</details>

---

### Validation Findings

Review mode with a machine-memory constraint — no local suite was run; disposition rests on the PR's existing CI plus the datadog top-level comment. Blocking build jobs (node/image builds, feature-unification, formatting/linting, unused-dependency) all pass, so the toolchain and workspace compile; the failures are test, metadata, and policy failures. Log retention had expired for the run, so the three suspected checks rest on retained step-summaries and are flagged for author confirmation rather than asserted as root-caused.

| # | Check | Severity |
|---|-------|----------|
| [1](https://github.com/midnightntwrk/midnight-node/pull/1800/checks) | Metadata Check — committed runtime metadata out of sync | Major |
| [2](https://github.com/midnightntwrk/midnight-node/pull/1800/checks) | check-changes — change file(s) missing the required PR link | Major |
| [3](https://github.com/midnightntwrk/midnight-node/pull/1800/checks) | Run tests — workspace suite fails (exit 100) | Critical |
| [4](https://github.com/midnightntwrk/midnight-node/pull/1800/checks) | Local Environment Tests — suspected change-induced (confirm) | Major |
| [5](https://github.com/midnightntwrk/midnight-node/pull/1800/checks) | Test Toolkit — suspected change-induced (confirm) | Major |
| [6](https://github.com/midnightntwrk/midnight-node/pull/1800/checks) | Toolkit E2E — suspected cascade of the toolkit failure (confirm) | Major |

<details>
<summary>Finding Details</summary>

#### VF-1. Metadata Check (Major, confirmed PR defect)
The committed runtime metadata is out of sync with freshly built metadata — the session-pallet rework changes the pallet and call surface, so metadata must be regenerated and committed. Deterministic, not flake.

#### VF-2. check-changes (Major, confirmed PR defect)
The change file(s) do not carry the required PR link that the repo changelog policy gates on. Deterministic, not flake.

#### VF-3. Run tests (Critical, confirmed PR defect)
The workspace test suite fails with exit 100 — a genuine test/compile failure, not infra. This is the suite carrying the rewritten one-epoch-lag rotation tests and the session-support tests. Change-induced.

#### VF-4 / VF-5 / VF-6. Local Environment Tests, Test Toolkit, Toolkit E2E (Major, suspected change-induced — confirm)
All three are consistent with the change: the local-env failure tracks the application-lag timing shift affecting finality and liveness; the toolkit failure plausibly follows the metadata change; the E2E fast-fail points to the toolkit image dependency rather than a timeout. None shows a flake signature, and none can be dismissed as infra on the retained evidence. Please confirm via a re-run or fresh-log read.

</details>

---

### Branch Hygiene

No findings. All 19 commits are GPG-signed and carry DCO sign-off trailers; the head-commit revert of the additional-session schedule is a deliberate, in-scope invariant restoration.

---

### Action Items

**Must Address (Blocking)**:
- [ ] Add an OnRuntimeUpgrade migration populating QueuedKeys / NextKeys / KeyOwner from pre-upgrade committee state; verify with try-runtime (CR-1, PF-6)
- [ ] Add a live-upgrade storage-migration test that fails without the migration (TR-1)
- [ ] Make key-registration failure non-silent — abort the rotation or validate all keys before applying any (CR-2, PF-9)
- [ ] Add a registration-failure rotation test; extend the mock to inject a set_keys failure (TR-2)
- [ ] Regenerate and commit runtime metadata (VF-1)
- [ ] Add the required PR link to the change file(s) (VF-2)
- [ ] Fix the failing workspace test suite (VF-3)
- [ ] Confirm or root-cause the three suspected CI failures via re-run / fresh logs (VF-4, VF-5, VF-6, PF-1)

**Should Address (Recommended)**:
- [ ] Correct the stale changelog line to describe the one-epoch-lag behaviour delivered at head (CR-3, PF-7)

**Could Address (Suggested)**:
- [ ] Add a dedup / duplicate-key drop test (TR-3)

**Nice to Have (Optional)**:
- [ ] Track the provider-count decrement fix to closure under [#1848](https://github.com/midnightntwrk/midnight-node/issues/1848) (CR-4)

---

*Posted by an automated review agent on behalf of @m2ux. The recommendation reflects an independent re-verification at head `4e48f8d`; the maintainers retain full discretion over disposition.*

File: partner-chains/toolkit/committee-selection/pallet/src/lib.rs
Comment (LGLO) [inline partner-chains/toolkit/committee-selection/pallet/src/lib.rs]: I'm wondering how much of it is still true and meaningful. Also, how much of this is assumed by BEEFy integration.

Comment (Klapeyron) [reply partner-chains/toolkit/committee-selection/pallet/src/lib.rs]: what about https://github.com/midnightntwrk/midnight-node/pull/1800/changes/456b3ea05542c44e116060b8ec88e8e978257568?

File: partner-chains/toolkit/committee-selection/pallet/src/tests.rs [hunk 7]
```diff
@@ -389,6 +439,16 @@ fn get_authority_round_robin_works() {
 			SessionCommitteeManagement::get_current_authority_round_robin(2),
 			Some(ALICE.ids_and_keys())
 		);
+		// A rotation queues the new committee; the current committee — the one the session
+		// machinery keeps active for this session — is unchanged.
+		assert!(rotate_committee().is_some());
+		assert_eq!(
+			SessionCommitteeManagement::get_current_authority_round_robin(0),
```
Comment (LGLO) [inline partner-chains/toolkit/committee-selection/pallet/src/tests.rs:446]: Conflicts with #1876

Comment (Klapeyron) [reply partner-chains/toolkit/committee-selection/pallet/src/tests.rs:446]: yes, I will address that in https://github.com/midnightntwrk/midnight-node/pull/1876 later (1876 deletes this test)

File: partner-chains/toolkit/committee-selection/pallet/src/pallet_session_support.rs [hunk 8]
```diff
@@ -1,68 +1,114 @@
...
+		let account_id = member.authority_id().into();
+
+		if keys_added.contains(&account_id) {
+			continue;
+		}
+
+		keys_added.insert(account_id.clone());
+		let keys = <T as pallet_session::Config>::Keys::from(member.authority_keys());
+		let call_result =
+			<pallet_session::Pallet<T> as SessionInterface>::set_keys(&account_id, keys);
```
Comment (LGLO) [inline partner-chains/toolkit/committee-selection/pallet/src/pallet_session_support.rs:89]: There is #1848 to fix it

**Review by LGLO — APPROVED** [review]
